### PR TITLE
Changed the model size calculation using `ByteCounter`

### LINKED ISSUE #10074: `get_model_size_mb` (`LightningModule.model_size`) shouldn't create temporary files in the current directory
Currently, the [`get_model_size_mb` function](https://github.com/PyTorchLightning/pytorch-lightning/blob/c3614f1c07caf2c1519fcbe37140f2aeef7a8308/pytorch_lightning/utilities/memory.py#L165) is implemented like this:
```python
def get_model_size_mb(model: Module) -> float:
    """Calculates the size of a Module in megabytes by saving the model to a temporary file and reading its size.
    The computation includes everything in the :meth:`~torch.nn.Module.state_dict`,
    i.e., by default the parameteters and buffers.
    Returns:
        Number of megabytes in the parameters of the input module.
    """
    # TODO: Implement a method without needing to download the model
    tmp_name = f"{uuid.uuid4().hex}.pt"
    torch.save(model.state_dict(), tmp_name)
    size_mb = os.path.getsize(tmp_name) / 1e6
    os.remove(tmp_name)
    return size_mb
```

This writes the model to disk and also introduces a race condition where the temporary file may be left behind if the user "Ctrl+C"s the program while the `torch.save` and `os.path.getsize` calls are running. I originally found this bug by investigating, what was creating the mysterious `cb0e0c0d62ee4c20986c67462c1105d3.pt` files in my project directory.

You can achieve the same result without touching the disk by doing
```python
import io

target = io.BytesIO()
torch.save(model.state_dict(), target)
size_mb = target.getbuffer().nbytes / 1e6
```

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #10074
all credits to @RuRo :)
Tests are already there for `get_model_size_mb`

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by rohitgr7** [comment]
@tchaton couldn't find a way to support this for Deepspeed/FSDP. Discussed this with @SeanNaren too and seems like its right not practically possible since weights are sharded and are not kept in a readable format. I'd suggest for now we can merge this for v1.5 and later we can explore more.

File: pytorch_lightning/utilities/memory.py [hunk 1]
```diff
@@ -25,6 +24,17 @@
...
 
 
+class ByteCounter:
+    def __init__(self) -> None:
+        self.nbytes: int = 0
+
+    def write(self, data: bytes) -> None:
+        self.nbytes += len(data)
+
+    def flush(self) -> None:
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/memory.py:34]:
```suggestion
    def write(self, data: bytes) -> None:
        self.nbytes += len(data)

    def flush(self) -> None:
```
Would like to some docstrings in these methods

**Review by kaushikb11 — APPROVED** [review]
LGTM! Could you add it to the class that it's not supported by DeepSpeed and Sharded plugins.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by SeanNaren — APPROVED** [review]
Looks good! can we add an issue around FSDP/DeepSpeed support for summarization of model size? As @tchaton mentioned it is possible, just will require some thought/implementation

File: CHANGELOG.md [hunk 2]
```diff
@@ -119,6 +119,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
     * Updated precision attributes in `DeepSpeedPlugin` ([#10164](https://github.com/PyTorchLightning/pytorch-lightning/pull/10164))
     * Added the ability to return a result from rank 0 in `DDPSpawnPlugin.spawn` ([#10162](https://github.com/PyTorchLightning/pytorch-lightning/pull/10162))
     * Added `pytorch_lightning.lite` package ([#10175](https://github.com/PyTorchLightning/pytorch-lightning/pull/10175))
+- Added `ByteCounter` to calculate model_size ([#10123](https://github.com/PyTorchLightning/pytorch-lightning/pull/10123))
```
Comment (awaelchli) [inline CHANGELOG.md:122]: I suggest to mention this in the changed section, that we are now using a byte counter instead of file io.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/memory.py [hunk 3]
```diff
@@ -25,6 +24,20 @@
...
+
+    def __init__(self) -> None:
+        self.nbytes: int = 0
+
+    def write(self, data: bytes) -> None:
+        """Stores the total bytes of the data."""
+        self.nbytes += len(data)
+
+    def flush(self) -> None:
+        pass
```
Comment (ananthsub) [inline pytorch_lightning/utilities/memory.py:38]: why do we need a byte counter class? the write/flush APIs aren't used anywhere

File: pytorch_lightning/utilities/memory.py [hunk 4]
```diff
@@ -25,6 +24,20 @@
 from pytorch_lightning.utilities.apply_func import apply_to_collection
 
 
+class ByteCounter:
```
Comment (carmocca) [inline pytorch_lightning/utilities/memory.py:27]: This class should be protected
